spec/imsc-hrm.html

```diff
@@ -471,6 +471,14 @@ <h2>Architecture</h2>
       "Intermediate Synchronic Document">Intermediate Synchronic Documents</a>, allowing glyphs to be copied into the Presentation
       Buffer instead of rendered, a more costly operation.</p>
 
+      <p class="note">The HRM permits a maximum rate of 12 <a>Intermediate Synchronic Documents</a> per second. This is
+      ultimately limited by the BDraw parameter and is intended to capture processing and presentation overhead. When
+      converting a [[CEA-608]] signal to IMSC, it is therefore impossible to create <a>IMSC Document Instances</a> that
+      generate an <a>Intermediate Synchronic Document</a> for every [[CEA-608]] packet, which are sampled at
+      approximately 59.94 Hz. It is instead preferable to coalesce sequences of [[CEA-608]] packets into longer
+      groupings, such as words, phrases, complete lines or paragraphs before creating an <a>IMSC Document Instance</a>,
+      and let the <a>presentation processor</a> perform any desired animation, e.g., typewriter effect.</p>
+
```

Comment:
I would suggest making it more generic than CEA-608. Isn't teletext having the same issue? Also, it seems to me that CEA-608 sampling depends on the video frame rate. If you have a 30fps video, your packets will be 33ms long. What about this suggestion?
```suggestion
      ultimately limited by the BDraw parameter and is intended to capture processing and presentation overhead. 
      It is therefore impossible to create HRM-valid <a>IMSC Document Instances</a> that
      generate an <a>Intermediate Synchronic Document</a> too frequently, such as for each new character.  
      For example, [[CEA-608]] data is typically carried in video frames. Depending on the video frame rate, 
      creating a new ISD from [[CEA-608]] data whenever a new frame is received would produce 
      HRM-valid documents. It is instead preferable to coalesce sequences of [[CEA-608]] packets into longer
```

Reply:
I would rather be specific about 608, and then suggest that the same constraints might apply to other systems.

Reply:
Happy to see an updated proposal.

Reply:
When converting a CEA-608 signal to IMSC, for example, it is therefore impossible to create IMSC documents that generate an ISD for every CEA-608 packet, which are sampled at approximately 59.94 Hz. It is instead preferable to coalesce sequences of CEA-608 packets into longer groupings, such as words, phrases, complete lines or paragraphs before creating an IMSC document, and let the presentation processor perform any desired animation, e.g., typewriter effect. The same approach can be applied to the conversion of other signals that transmit subtitles and captions character by character at the video sampling rate.

Reply:
I think the choice between "being generic and then giving an example" vs. "being specific and then generalizing" is a matter of taste. In this case, I'd prefer the former, but if others prefer the latter, I'm fine with that.

It is not that it's "impossible to create IMSC documents", it's more that it would create HRM-invalid document. I tried to fix that in my suggestion. I think it's important to be precise.
In my experience, 608 sampling is not necessarily 59.94Hz. It's frame rate specific. I also tried to fix that in my suggestion.

Reply:
> In my experience, 608 sampling is not necessarily 59.94Hz. It's frame rate specific. I also tried to fix that in my suggestion.

This is news to me. CTA 608 is defined in the context of line 21 services, which, AFAIK, is only defined for 59.94 field rate (see Figure 2 at CTA 608) . Where did you see CTA 608 used at different field/frame rate?

Reply:
> This is news to me. CTA 608 is defined in the context of line 21 services, which, AFAIK, is only defined for 59.94 field rate (see Figure 2 at CTA 608) . Where did you see CTA 608 used at different field/frame rate?

Commercial video encoders can embed 608 captions at any frame rate. They are embedded in SEI messages. One message per frame. I have example files at 30fps for example.

Reply:
> The same approach can be applied to the conversion of other signals that transmit subtitles and captions character by character at the video sampling rate.

Are there any others that do this?

Reply:
>Commercial video encoders can embed 608 captions at any frame rate. They are embedded in SEI messages. One message per frame. I have example files at 30fps for example.

The 608 packets are still sampled at 59.94 Hz, but grouped into 30 fps SEI messages, right?

Reply:
> Are there any others that do this?

I do not know, and this is why I suggest we keep the example to 608 unless someone has a concrete other example.

Reply:
> The 608 packets are still sampled at 59.94 Hz, but grouped into 30 fps SEI messages, right?

[This document](https://download.tek.com/document/2PW-24267-0.pdf) (I don't know how truthful it is) says:
>  In the ATSC (Advanced Television Systems Committee) / MPEG
encoding process the frame rate of the video signal should match the frame rate of
the closed caption data stream otherwise this can potential cause problems in the
encoding process and may prevent captions from being transmitted.

It also contains the following table:
![image](https://github.com/w3c/imsc-hrm/assets/1830314/ee631748-2cdc-45c2-9ffd-fea14293199f)

The fact that lower frame rates can carry more bytes could indeed be interpreted as a source sampling at a higher frame rate as you suggest. In my experience, even at 24 or 30fps, you don't get twice the rate.

My suggestion was to remove the wording around 59.94Hz and simply talk about caption rate being related to the video frame rate.

Reply:
> Are there any others that do this?

Doesn't Teletext behave similarly? [This page](https://support.telestream.net/s/article/Encoding-OP-47-Teletext) (also not sure how truthful it is) indicates:
> the video frame rate must be 25 fps or 50 fps, because Teletext only supports these frame rates.

Reply:
> My suggestion was to remove the wording around 59.94Hz and simply talk about caption rate being related to the video frame rate.

I do see the advantage in being vague when we have a concrete example.

Reply:
> Doesn't Teletext behave similarly?

It encodes a lot more information on each frame: rather than having the text dribble in character by character, a whole 23x40 character grid can be delivered (it's about 1kb) every frame. It's the small number of characters per field aspect of 608 that causes the issue here, not the frame-bound nature of the delivery mechanism.

Reply:
> It encodes a lot more information on each frame: rather than having the text dribble in character by character, a whole 23x40 character grid can be delivered (it's about 1kb) every frame.

Thanks.

> It's the small number of characters per field aspect of 608 that causes the issue here, not the frame-bound nature of the delivery mechanism.

I disagree. The problem is the refresh rate being to high, which comes from the frame-bound delivery. If teletext could deliver even more, it would be more reasons for the HRM validation to fail.

Reply:
> > My suggestion was to remove the wording around 59.94Hz and simply talk about caption rate being related to the video frame rate.
> 
> I do see the advantage in being vague when we have a concrete example.

Do you have a pointer that indicates 608 is always sampled 59.94Hz? If not, I would remove that part.

Reply:
>I disagree. The problem is the refresh rate being to high, which comes from the frame-bound delivery. If teletext could deliver even more, it would be more reasons for the HRM validation to fail.

Ah, I'm seeing this differently: since Teletext essentially delivers a whole "page" in one go, there's no scope for a naive implementation to build up that page over several timed iterations. Whereas with 608, where each field carries a delta, the naive implementation simply appends those new characters and issues a new ISD at the field time, even though that's nobody's proper understanding of how 608 is supposed to work or be used. It's this naive field-by-field behaviour that was reported as being the cause of the content validation failure when analysed with the HRM.